[BEAM-3181][Nexmark][SQL] Implement query0

akedin opened the pull request:
This implements a basic pass-through query0. 

This PR does not depend on the refactoring PRs and can be merged independently while the other ones are being worked on.

This sets up a basic scaffolding which will help implementing other queries.

[BEAM-3181](https://issues.apache.org/jira/browse/BEAM-3181) is a subtask of [BEAM-3147](https://issues.apache.org/jira/browse/BEAM-3147)


Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

akedin commented:
@kennknowles , @xumingming , @iemejia , can you take a look at this one?

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
FAILURE
 
--none--

echauchot commented:
@akedin if you don't mind, I'd like to take a look as well.
R: @echauchot

kennknowles commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/Auction.java:
```diff
@@ -80,14 +80,14 @@ public Auction decode(
   /** Extra auction properties. */
   @JsonProperty
-  private final String itemName;
+  public final String itemName;
```
I mentioned `@AutoValue` before and I think you can do this, at least for Jackson:

```java
@AutoValue
@AutoValue.CopyAnnotations
public abstract class Auction {
  @JsonProperty
  public abstract String getItemName();
}
```

I don't know that it is possible to use `@AutoValue` with avro schema generation, though.

akedin replied:
<!--thread_id:cc_150930883_t; commit:5d81ac0253726b036b7f1264e3719ee7507fedf7; resolved:0-->
<!--section:context-quote-->
> **kennknowles** wrote:
> I mentioned `@AutoValue` before and I think you can do this, at least for Jackson:
> 
> ```java
> @AutoValue
> @AutoValue.CopyAnnotations
> public abstract class Auction {
>   @JsonProperty
>   public abstract String getItemName();
> }
> ```
> 
> I don't know that it is possible to use `@AutoValue` with avro schema generation, though.

<!--section:body-->
I don't see a straightforward way to do this:
 - Avro seems to require a no-args constructor in our case, i.e. we seem to end up here when creating records instances: https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java#L351 
 - AutoValue doesn't support generation of no-args constructor, from what I can see;

Probably we could work around this by having Avro-specific subclasses of the model classes, but I don't think it should be done as part of this effort.

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery0.java:
```diff
@@ -0,0 +1,68 @@
+ * <p>{@link Bid} events are used here at the moment, ås they are most numerous
+ * with default configuration.
+ */
+public class SqlQuery0 extends PTransform<PCollection<Event>, PCollection<BeamRecord>> {
```
<!--thread_id:cc_150883553_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Part of the job of original Query0 was to mesure the impact of serialization and metrics to serve as a baseline. SqlQuery0 measures only SQL impact. If the user were to compare SqlQuery0 to Query0 he would compare different things. Is this possible to add `Metrics.counter` and encode/decode roundtrip to SqlQuery0?

<!--section:body-->
I agree that it tests a different thing, but I cannot think of a good way to test exactly same stuff in SQL. 

Java Query0 transform takes an element, runs it though the coder and returns with no change. 

One way to measure the serialization in SQL version would be actually to avoid SQL query execution completely. Instead a version of the existing Java Query 0 can be executed against a PCollection of BeamRecords. This way we can measure the coders for BeamRecords. While this can be a better way of measuring coders impact I am not convinced that this can provide a useful baseline. For example, I don't think a user of SQL can control, or even assume how many of such serializations will be done by any actual query. It's hard to reason about what happens internally in projections, joins, udf calls. Or things like type conversions/casts that happen internally between Beam/Calcite can change the way data is treated.

Another way is to execute an actual SQL query. But then it becomes hard to add arbitrary transform or monitoring code into the query, you can only use what SQL allows to express. You can add a UDF with an arbitrary code, but this again will be testing a different thing from Java Query0, as it doesn't even operate on the same BeamRecord as the input or output of the query.

Do you have an opinion on what approach will be better?

echauchot replied:
Indeed, Query0 and SQLQuery0 are two different things almost impossible to compare their serialization, it's pointless to try to. My point was mainly to add the same Metrics call to the SQL queries. SqlQuery0 could serve as a baseline for the SQL part of nexmark comparing it only to other SQL versions of the queries.

echauchot replied:
Can you add metrics call to the sql query0?

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapters.java:
```diff
@@ -0,0 +1,125 @@
+/**
+ * Maps Java model classes to Beam SQL record types.
+ */
+public class ModelFieldsAdapters {
```
<!--thread_id:cc_151126021_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Generally speaking, I'm not a big fan of `ModelFieldsAdapter` and `ModelFieldsAdapters` classes. I understand that you have concrete adapters for nexmark model objects and `ModelFieldsAdapters.ADAPTERS` singleton to have all the adapters ready for use to convert in `ToBeamRecord` and also to get the coders. But IMHO I think it would be simpler and more object oriented to have the mapper/converter methods directly in Bid/Auction/Person... classes. WDYT? Is there something that I miss that prevents it?

<!--section:body-->
I see how it can be confusing. My reasonsing went something like this:

 - separation of concerns / data from algorithms. Stream of events produced by Generator is just data which you can apply algorithms on top of. Specific queries are implementation details of such algorithms. They are not limited to SQL. It's only current implementation details of SQL queries that requires objects to be converted to BeamRecords. So it feels that objects themselves should not be exposed to the fact that they are used by a specific kind of algorithms. For one thing, nothing stops Beam SQL to eventually be able to query objects directly without explicit in-advance conversion;

 - less coupling in general. Current Java models and generators have nothing to do with SQL. They can be used unchanged in any Beam-related code. For example I can see something like this used for integration testing. SQL representation of the models, on the other hand, can, and likely will change;

 - single responsibility. Current Java models are basically just serializable pojos. These utility classes also do one thing each. I see no reason to add behavior to pojos;

 - from what it looks like, not all models require conversion for use in SQL queries, only the ones being produced by the Generator. Query outputs are fine being BeamRecords, from what I can see so far;

 - my hope is that eventually this conversion layer will be gotten rid of, so that data is generated in a manner directly consumable by SQL. For example by making SQL understand Pojos, or by generating BeamRecords directly;


I hope this makes sense.

echauchot replied:
Yes it makes a lot of sense, your aim is much clearer now. I agree with the low coupling and with the SRP too (even if there is not a lot of SRP in beam ;) ). Also, considering that the SQL converters might become useless in the future, extracting that conversion logic is a good thing indeed. The only conversion methods that belong to the model objects will be the encode/decode that are very unlikely to disappear in the future :)

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/BeamRecordSqlTypeBuilder.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Helper class to construct {@link BeamRecordSqlType}.
+ */
+public class BeamRecordSqlTypeBuilder {
```
<!--thread_id:cc_150889287_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> This class seems to have nothing related to nexmark it should be in beam.sdk.extensions.sql package I think. But I'm not very sure this class is needed as it avoids only a few lines of code. WDYT?

<!--section:body-->
I wasn't sure whether it makes sense to add it to the main SQL package at this moment, don't know how useful it is for wider applications. I see no problems moving it though, if you think it belongs there.

Main reason I made it is for readability. Currently `BeamRecordSqlType.create()` only accepts separate lists of names and types, and it's hard to match them between each other, [like here](https://github.com/apache/beam/blob/39e66e953b0f8e16435acb038cad364acf2b3a57/sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlBuiltinFunctionsIntegrationTestBase.java#L65-L71). This class makes it clear which field has which type.

echauchot replied:
Good point on the matching of the arguments. I think this kind of readability improvement could profit all the users of SQL DSL, might me worth a separate (because unrelated to nexmark) PR  toward SQL DSL. Have you considered putting the builder directly in the `BeamRecordSqlType` a bit like the `Autovalue` pattern for example?

akedin replied:
Sure, I'll move it to the main SQL package in a separate PR

akedin replied:
Implemented the builder in the main SQL package: https://github.com/apache/beam/pull/4168

akedin commented on sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdaptersTest.java:
```diff
@@ -0,0 +1,161 @@
+      new Person(3L, "name", "email", "cc", "city", "state", 329823L, "extra");
+
+  private static final BeamRecordSqlType PERSON_RECORD_TYPE = BeamRecordSqlTypeBuilder.of()
+      .withLongField("id")
```
<!--thread_id:cc_151145780_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Same comment as above: WDYT about doing this logic inside the model object to avoid spreading the knowledge of the fields of `Person` object outside of the `Person` class?

<!--section:body-->
I considered it but for the reasons in my other comment I decided against it.

This knowledge is essential for the SQL queries though. Queries need to know how to get data from records. These are just utility classes which determine this field mapping for specific set of SQL queries.

akedin commented on sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapterTest.java:
```diff
@@ -0,0 +1,56 @@
+
+    ModelFieldsAdapter<String> adapter = new ConcreteAdapter(recordType);
+
+    assertSame(recordType, adapter.getRecordType());
```
<!--thread_id:cc_151144510_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> I don't understand what it tests: the only thing that I see it testing is that `recordType` field is properly assigned in `ConcreteAdapter` 's superclass constructor. Did I miss something?

<!--section:body-->
That is pretty much what happens. It tests that record type is assigned and returned as expected.

echauchot replied:
Then, I don't think it is very useful to test line   `ModelFieldsAdapter(BeamRecordSqlType recordType) {this.recordType = recordType; }` of ModelFieldsAdapter class.

echauchot replied:
IMHO, I would remove this test

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapters.java:
```diff
@@ -0,0 +1,125 @@
+    return new ModelFieldsAdapter<Person>(
+        BeamRecordSqlTypeBuilder.of()
+            .withLongField("id")
+            .withStringField("name")
```
<!--thread_id:cc_151126910_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Related to my last comment: having the adapter logic in the `Person` class would avoid spreading the knowledge of the `Person` fields outside the `Person` class.

<!--section:body-->
I considered it but for the reasons in my other comment I decided against it.

This knowledge is essential for the SQL queries though. Queries need to know how to get data from records. These are just utility classes which determine this field mapping.

echauchot replied:
Separating converter from pojo is a good thing for coupling as we discussed. But if pojo change we will have to change the code in plenty other places, it's less good for encapsulation. There is still a way to keep the field knowledge inside the pojo and keep the converter outside the pojo. Have you considered using reflexion for that? But I don't know it it's worth it. WDYT?

akedin replied:
I am not sure I understand how we can keep the field knowledge inside the Pojos but conversion logic outside, could you share an example code snippet?

How I see it:

Main problem is that to construct a BeamRecord you need the Pojo field values in a speciofic order, matching the order of the fields in the record type. Record type (and the order of its fields) is defined by what specific SQL query expects as an input. This means that Pojo fields retrieval has to be done at the moment of the actual conversion, when we have both the Pojo instance and the record type. 

So now the question is reduced to how to get the Pojo field values in the correct order. It can probably be done with something like this at conversion time (pseudocode):

```java
public BeamRecord pojoToRecord(T pojo) {
  BeamRecordSqlType recordType = getRecordTypeForPojo(pojo);

  List<Object> fieldValues = getFieldNames(recordType)
       .map(fieldName -> getFieldValue(pojo))
       .collect(toList());

  return new BeamRecord(recordType, fieldValues);
}
```

Now the question becomes how you do `getFieldValue(pojo, fieldName)`. 
Doing it with reflection at runtime or code generation is one way. This is probably the right way but it feels like it's out of scope here.

Another way would be to define a mapping somewhere, like this (pseudocode):
```java
class Person {
  Map<String, Supplier<Object> FIELD_VALUES_SUPPLIERS = ImmutableMap.of(
     "name", Person::getName,
     "address", Person::getAddress);

  public String getName() { ... }
  public String getAddress() { ... }
```
Something along these lines could be generic enough to keep in Pojo. Problem with this is I don't know how to express this in a readable way in Java7.

Instead of the above approaches I decided to declare the record types and Pojo field retrieval logic in one place, so that at least:
 - it's easy to see how specific Pojo fields map to specific record fields;
 - it's relatively easy to see and debug if there's a mismatch between types or field names;
 - it's easy to argue about performance of this piece;
 - doesn't require a lot of complicated code in the models or infrastructure;

echauchot replied:
You're right it is out of the scope of this PR. all my point was to make the fields mapping generic enough to be included in sql package (like you extracted the other feature) rather than coding it for each pojo. Like you said above implementing `getFieldValue(pojo, fieldName)` using reflection might be a good thing if it is feasible.

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/ToBeamRecord.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Convert Java model object to BeamRecord.
+ */
+public class ToBeamRecord {
```
<!--thread_id:cc_151138782_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Same as comment about the adapters classes, please read it before. WDYT about including this mechanism it to the Person/Bid/Auction ... classes? Though It might entail the need to create a superclass of the model classes to deduplicate code.

<!--section:body-->
I personally don't think that this would be a good application of inheritance. I think that those data objects are not inherently supposed to be able to convert themselves to BeamRecords, and not all of them might even need to support it, so having an external utility class makes more sense to me.

echauchot replied:
ok, fair enough

kennknowles commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java:
```diff
@@ -1154,4 +1140,81 @@ public NexmarkPerf run(NexmarkConfiguration runConfiguration) {
+  }
+
+  private List<NexmarkQueryModel> createJavaQueryModels() {
+    return Arrays.asList(
```
<!--new_thread; commit:964cc527e15c17e26b9584aaaf993cddc225398b; resolved:1-->
Nit: `Immutable.list.of(... elems ...)` is a little more readable than `Arrays.asList`.

akedin replied:
<!--thread_id:cc_154870916_t; commit:964cc527e15c17e26b9584aaaf993cddc225398b; resolved:1-->
<!--section:context-quote-->
> **kennknowles** wrote:
> Nit: `Immutable.list.of(... elems ...)` is a little more readable than `Arrays.asList`.

<!--section:body-->
I agree but `ImmutableList.of()` doesn't allow `null`s. There is an explicit logic to skip query models if they are `null`, so I will keep it for now. Refactoring that bit is out of scope of this PR.

kennknowles commented on sdks/java/nexmark/pom.xml:
```diff
@@ -25,6 +25,10 @@
   </parent>
 
+  <properties>
+    <calcite.version>1.13.0</calcite.version>
```
<!--new_thread; commit:964cc527e15c17e26b9584aaaf993cddc225398b; resolved:0-->
This variable is not referenced here. The version in sdks/java/extendsion/sql/pom.xml is already this. So is it already solved by a prior PR? Or did I just not follow the flow of this variable?

akedin replied:
<!--thread_id:cc_154870918_t; commit:964cc527e15c17e26b9584aaaf993cddc225398b; resolved:1-->
<!--section:context-quote-->
> **kennknowles** wrote:
> This variable is not referenced here. The version in sdks/java/extendsion/sql/pom.xml is already this. So is it already solved by a prior PR? Or did I just not follow the flow of this variable?

<!--section:body-->
Forgot to clean this up. Removed.

kennknowles commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/BeamRecordSize.java:
```diff
@@ -0,0 +1,136 @@
+import org.apache.beam.sdk.values.BeamRecord;
+
+/**
+ * KnownSize implementation to estimate the size of a BeamRecord, similar to Java model.
```
<!--new_thread; commit:dab81baf15038727942d0f963bd402793a80ec30; resolved:1-->
Nit: Could use {@link Knownsize} and {@link BeamRecord}.

akedin replied:
<!--thread_id:cc_154870922_t; commit:dab81baf15038727942d0f963bd402793a80ec30; resolved:1-->
<!--section:context-quote-->
> **kennknowles** wrote:
> Nit: Could use {@link Knownsize} and {@link BeamRecord}.

<!--section:body-->
done

kennknowles commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/BeamRecordSize.java:
```diff
@@ -0,0 +1,136 @@
+ * <p>The BeamRecord size is calculated at creation time.
+ *
+ * <p>Field sizes are sizes of Java types described in {@link BeamRecordSqlType}. Except strings,
+ * which are assumed to be taking 1-byte per character plus 1 byte size.
```
<!--new_thread; commit:dab81baf15038727942d0f963bd402793a80ec30; resolved:1-->
This is fine for an estimate of the benchmark as long as it has no operational significance (yet).

akedin replied:
<!--thread_id:cc_154870923_t; commit:dab81baf15038727942d0f963bd402793a80ec30; resolved:1-->
<!--section:context-quote-->
> **kennknowles** wrote:
> This is fine for an estimate of the benchmark as long as it has no operational significance (yet).

<!--section:body-->
I am not sure how usable or valuable these metrics are at the moment. 

It's the same assumption current java model has. It's based on what the coder does for the kinds of strings we have here.

kennknowles commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/Person.java:
```diff
@@ -75,10 +75,10 @@ public Person decode(InputStream inStream)
 
   @JsonProperty
-  private final String creditCard;
+  public final String creditCard;
```
<!--new_thread; commit:84f538f392fa0a3244e7ffe4819adbef86d6e289; resolved:1-->
Something about this is funny ;-)

echauchot replied:
:)

echauchot commented:
LGTM  modulo 2 minor things: the metrics call and the removal of the test

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery0.java:
```diff
@@ -0,0 +1,68 @@
+ * <p>{@link Bid} events are used here at the moment, ås they are most numerous
+ * with default configuration.
+ */
+public class SqlQuery0 extends PTransform<PCollection<Event>, PCollection<BeamRecord>> {
```
<!--thread_id:cc_150883553_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:1-->
<!--section:context-quote-->
> **echauchot** wrote:
> Can you add metrics call to the sql query0?

<!--section:body-->
Done

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapters.java:
```diff
@@ -0,0 +1,125 @@
+/**
+ * Maps Java model classes to Beam SQL record types.
+ */
+public class ModelFieldsAdapters {
```
<!--thread_id:cc_151126021_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:1-->
<!--section:context-quote-->
> **echauchot** wrote:
> Yes it makes a lot of sense, your aim is much clearer now. I agree with the low coupling and with the SRP too (even if there is not a lot of SRP in beam ;) ). Also, considering that the SQL converters might become useless in the future, extracting that conversion logic is a good thing indeed. The only conversion methods that belong to the model objects will be the encode/decode that are very unlikely to disappear in the future :). Can you just rename the names because they are so close to each others that it is a bit confusing.

<!--section:body-->
Renamed

akedin commented on sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapterTest.java:
```diff
@@ -0,0 +1,56 @@
+
+    ModelFieldsAdapter<String> adapter = new ConcreteAdapter(recordType);
+
+    assertSame(recordType, adapter.getRecordType());
```
<!--thread_id:cc_151144510_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:1-->
<!--section:context-quote-->
> **echauchot** wrote:
> IMHO, I would remove this test

<!--section:body-->
removed

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapters.java:
```diff
@@ -0,0 +1,125 @@
+    return new ModelFieldsAdapter<Person>(
+        BeamRecordSqlTypeBuilder.of()
+            .withLongField("id")
+            .withStringField("name")
```
<!--thread_id:cc_151126910_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:1-->
<!--section:context-quote-->
> **echauchot** wrote:
> You're right it is out of the scope of this PR. all my point was to make the fields mapping generic enough to be included in sql package (like you extracted the other feature) rather than coding it for each pojo. Like you said above implementing `getFieldValue(pojo, fieldName)` using reflection might be a good thing if it is feasible.

<!--section:body-->
This discussion made me start implementation of [automatic BeamRecordType generation](https://github.com/apache/beam/pull/4204). Keeping these classes as they are right now until record type generation is ready.

echauchot replied:
+1

echauchot commented:
retest this please

echauchot commented:
once again LGTM
@akedin : can you please: rebase on master, squash commits into one which message would be something like 

> [BEAM-3181][Nexmark][SQL] Implement query0
Also make sure the tests pass. I'll merge it once it is done

akedin commented:
> @akedin : can you please: rebase on master, squash commits into one which message would be something like
> [BEAM-3181][Nexmark][SQL] Implement query0

Rebased, squashed, ran tests locally

echauchot commented:
retest this please

jbonofre commented:
Jenkins job is stuck due to Dataflow runner. Let me take a look.

akedin commented:
I am working on fixing the shadow plugin configuration in the gradle build

akedin commented:
rebased, updated the gradle build config

akedin commented:
retest this please

akedin commented:
retest this please

echauchot commented:
Well, wordcount examples are still failing but they are unrelated to the changes in this PR. I'll merge it

echauchot commented:
the jenkins build is stuck on wordcount tests on dataflow runner. I tested the build locally but it fails on `runners.core.construction.ArtifactServiceStagerTest`  and `sdk.io.kinesis.KinesisMockReadTest` which are known flaky tests and are unrelated to this PR. The nexmark module build passes. So I think we can merge this PR

kennknowles commented:
Agreed.
